text/1013-datetime-arithmetic.rst

```diff
@@ -0,0 +1,226 @@
+be an ``int64`` representing the number of days. It is unambiguous an often
+what is practicaly useful. This logic may be extended to also include a ``+``
+operator between ``local_date`` and ``int64`` so that flipping though this
+virtual calendar is simplified and doesn't require bulkier
+``relative_duration`` for many common use-cases.
+
+In the future we want to expose ``generate_series`` in some fashion (likely
+via range types). Using an ``int64`` step for ``local_date`` makes it easier
+to produce a sequence of dates than using a ``relative_duration`` step for a
+lot of common use-cases.
+
+Working with ``relative_duration`` sometimes requires converting between hours
+and days or between days and months. We will need to expose the ``justify``
+functionality to achieve this.
+
+
+Specification
+=============
+
+The ``local_datetime`` will introduce the following subtraction operator::
+
+  CREATE INFIX OPERATOR
+  std::`-` (
+    l: cal::local_datetime,
+    r: cal::local_datetime
+  ) -> cal::relative_duration {
+      USING SQL $$
+          SELECT ("l" - "r")::edgedb.relative_duration_t
+      $$;
+  };
+
+The result is expected to be a ``relative_duration`` in days and
+hours/minutes/seconds. This is the most precise format as ``local_datetime``
+cannot account for DST and so 1 day and 24 hours are always equivalent
+durations in this context.
+
+The ``local_time`` will introduce the following subtraction operator::
+
+  CREATE INFIX OPERATOR
+  std::`-` (
+    l: cal::local_time,
+    r: cal::local_time
+  ) -> cal::relative_duration {
+      USING SQL $$
+          SELECT ("l" - "r")::edgedb.relative_duration_t
+      $$;
+  };
+
+The result is expected to be a ``relative_duration`` in the range [-24 hrs, 24
+hrs].
+
+The ``local_date`` will introduce the following subtraction operator::
+
+  CREATE INFIX OPERATOR
+  std::`-` (
+    l: cal::local_date,
+    r: cal::local_date
+  ) -> int64 {
+      USING SQL $$
+          SELECT ("l" - "r")::int8
+      $$;
+  };
+
+The result is expected to be the number of days.
+
+
+Duration functions
+------------------
+
+We also will introduce normalization functions for ``relative_duration``::
+
+CREATE FUNCTION
+cal::normalize_relative_hours(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_hours';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_days(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_days';
+};
+
+CREATE FUNCTION
+cal::normalize_relative_duration(dur: cal::relative_duration)
+  -> cal::relative_duration
+{
+    USING SQL FUNCTION 'justify_interval';
+};
+
+The ``normalize_relative_hours`` converts 24 hrs chunks into days.
+The ``normalize_relative_days`` converts days into months assuming that 1 month = 30 days.
+The ``normalize_relative_duration`` is the equivalent of calling
+``normalize_relative_days(normalize_relative_hours(x))``.
```

Comment:
I actually think that the original "justify" name is probably better here, though whatever we choose it's not going to be 100% descriptive unless we go the super-verbose route.

Reply:
> though whatever we choose it's not going to be 100% descriptive unless we go the super-verbose route.

My problem is mostly that `normalize_relative_days` is not clear whether it normalizes up (i.e. days to months and years) or down (i.e. hours to days). Also technically `normalize_relative_hours` normalizes "microseconds" instead (i.e. there is no `hours` counter in the internal structure).

Reply:
Also, I like the precedent that [systemd set](https://www.freedesktop.org/software/systemd/man/systemd.time.html) is that months are counted as `30.44` days and years are set to `365.25` days. For big durations, it's much more accurate for human presentation.

On the other hand, "normalizing" relative datetime that had just days and getting non-zero hours in the "normalized" variant is also unexpected for most users. So it's more useful for creating rather than formatting and for durations rather than relative durations.

And then, as you have said, years are also useful and they will be calculated from months, but their duration will be `360` days with the semantics defined in the current spec.

What I'm trying to say is that there is no generally agreed "normalization" operation. 

So we might want few formatting and/or truncation operations instead, with customizable ways to convert units.

Reply:
I've added an overload for `duration_truncate` for `relative_duration` to this PR. So with some combination of `duration_get`, `duration_truncate` and possibly some normalizations you can try to massage the `relative_duration` into a reasonable presentation.

At the end of the day, if you're trying to normalize `relative_duration` you've already lost your reference point and will inevitably make some approximations, so that process is like calling `round`. For more accurate `relative_duration` values we need to expose `age` (possibly under some other name), which performs the arithmetic as accurately as possible for the given 2 datetimes.

Reply:
> if you're trying to normalize `relative_duration` you've already lost your reference point and will inevitably make some approximations, so that process is like calling `round`.

Yes. But there is a difference. It's fine if `date + normalize_relative_duration(x)`
comparing to `date + x` is off by hours (or maybe days on large numbers). But with the semantics above, it will be off by two years if `x > "55000 days"`.

I you really want something like rounding behavior then month length should be about `30.44` days and rounding is used to throw away fractional days. But I'm not sure if this makes sense if fractional days are already present in the input.